// @iopipe/event-info/dist/plugins/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sqs = exports.sns = exports.slsIntegrationLambda = exports.scheduled = exports.s3 = exports.kinesis = exports.firehose = exports.cloudFront = exports.apiGateway = exports.alexaSkill = undefined;

var _alexaSkill = require('./alexaSkill');

var alexaSkill = _interopRequireWildcard(_alexaSkill);

var _apiGateway = require('./apiGateway');

var apiGateway = _interopRequireWildcard(_apiGateway);

var _cloudFront = require('./cloudFront');

var cloudFront = _interopRequireWildcard(_cloudFront);

var _firehose = require('./firehose');

var firehose = _interopRequireWildcard(_firehose);

var _kinesis = require('./kinesis');

var kinesis = _interopRequireWildcard(_kinesis);

var _s = require('./s3');

var s3 = _interopRequireWildcard(_s);

var _scheduled = require('./scheduled');

var scheduled = _interopRequireWildcard(_scheduled);

var _slsIntegrationLambda = require('./slsIntegrationLambda');

var slsIntegrationLambda = _interopRequireWildcard(_slsIntegrationLambda);

var _sns = require('./sns');

var sns = _interopRequireWildcard(_sns);

var _sqs = require('./sqs');

var sqs = _interopRequireWildcard(_sqs);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.alexaSkill = alexaSkill;
exports.apiGateway = apiGateway;
exports.cloudFront = cloudFront;
exports.firehose = firehose;
exports.kinesis = kinesis;
exports.s3 = s3;
exports.scheduled = scheduled;
exports.slsIntegrationLambda = slsIntegrationLambda;
exports.sns = sns;
exports.sqs = sqs;
